music21/stream/__init__.py

```diff
@@ -6367,36 +6376,21 @@ def updateEndMatch(n):
                 posConnected = []  # reset to empty
 
         # all results have been processed
-        # presently removing from notes, not resultObj, as index positions
-        # in result object may not be the same
         posDelete.reverse()  # start from highest and go down
 
+        for i in posDelete:
+            # environLocal.printDebug(['removing note', notes[i]])
+            # get the obj ref
+            nTarget = notes[i]
+            # Recurse rather than depend on the containers being Measures
+            # https://github.com/cuthbertLab/music21/issues/266
+            returnObj.remove(nTarget, recurse=True)
+        returnObj.coreElementsChanged()
+
         if retainContainers:
-            for i in posDelete:
-                # environLocal.printDebug(['removing note', notes[i]])
-                # get the obj ref
-                nTarget = notes[i]
-                # go through each container and find the note to delete
-                # note: this assumes the container is Measure
-                # TODO: this is where we need a recursive container Generator out
-                for sub in reversed(returnObj.getElementsByClass('Measure')):
-                    try:
-                        i = sub.index(nTarget)
-                    except StreamException:
-                        continue  # pass if not in Stream
-                    junk = sub.pop(i)
-                    # get a new note
-                    # we should not continue searching Measures
-                    break
-            returnObj.coreElementsChanged()
             return returnObj
-
         else:
-            for i in posDelete:
-                # removing the note from notes
-                junk = notes.pop(i)
-            notes.coreElementsChanged()
-            return notes
+            return returnObj.flat.notesAndRests.stream()
```

Comment:
WOW.  and that's it!  :-)

Reply:
Actually, check me on this -- why are we doing `.notesAndRests.stream()` ?  I would have thought that StripTies with retainContainers=False returned a Stream that was identical to the original flat stream except that the tied notes have been removed.  I wouldn't want the Clefs, etc. removed -- is that what we've been doing for a long time?  Uh oh.  I think just `return returnObj` should be good, and people can do `.notesAndRests` themselves.

if `retainContainers` default was True, then I would remove it completely as an option, since `.flat` is simple to call later, but since it exists with a True default, we should keep that.

Reply:
Yep, that was the existing behavior for `retainContainers=False`, strip clefs, everything--however, the complex defaults made it difficult to reach. The keyword supplied or the default value from the method signature is _ignored_ if the input stream has part-like substreams. So all scores are forced to run under `retainContainers=True`. Otherwise, we read the keyword, which defaults False but a user can set True. Nasty to explain in the docs.

Everywhere else in the m21 codebase sets `retainContainers=True`. (Well, now that I commented out those two lines in the two feature extractors that were causing trouble.) There is one example in the user's guide that uses `retainContainers=False`:
```
One thing to notice is that the note E extends now beyond the end of the first 2/4 measure. 
The second D, in measure 2, by contrast, does not begin at the beginning of the measure, 
but instead halfway through the first measure. This is why it's sometimes most helpful 
to run stripTies() without retaining containers:

stripped2 = littlePiece.stripTies(retainContainers=False)
stripped2.show('text', addEndTimes=True)
{0.0 - 1.5} <music21.note.Note D>
{1.5 - 3.0} <music21.note.Note E>
{3.0 - 4.5} <music21.note.Note D>
{4.5 - 6.0} <music21.note.Note F>
```

I think it would be a big win to remove the keyword. Single responsibility principle. No difficult to explain keyword-overwriting.

Reply:
It's curious, even some of the calls that pass no keywords to `stripTies` just end up calling `.notes` later anyway, probably because of the fickle keyword behavior:
https://github.com/cuthbertLab/music21/blob/6e5b01d1852097c6da94e4cbb492c36105d49902/music21/analysis/discrete.py#L1287

Reply:
Actually, now that I look at this, the only reason I derived the return object this way was because a few commits ago I was trying to run `voicesToParts` and indented most of the algorithm under that, but now I don't have to re-derive the return object, I can just return `notes` from near the top of the function.

Reply:
Nope, doesn't work. Still need to re-derive it because I'm only removing notes from `returnObj` now. Re-deriving at the end of the function is extra overhead -- this all points to removing the keyword, I think. Deprecating the keyword and rewriting some docs could be handled in another PR down the line, no?

Reply:
I would say, let's do this -- remove the `.notesAndRests` and make a note saying that non-Notes are retained now.  And change retainContainers to default to True, and put a note saying that retainContainers has defaults to `True` now, and that the keyword will be removed completely in a future version around 2022 -- this gives people who need to support both music21 v5 and v6 a way of keeping the containers (via `retainContainers=True`) but gives two years as a transition time.

I think that this is the best way to help people support v5 and v6 and future versions, but not to commit to keeping retainContainers forever!